### PR TITLE
update for titiler-pgstac 0.1.0a3

### PR DESCRIPTION
**Pull request by vincentsarago**
## Description

This PR update TiTiler-PgSTAC requirement (+titiler and rio-tiler)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

TODO 

Locally, just made 
## Checklist:

Please delete options that are not relevant.

- [ ] I have performed a self-review
- [ ] Changelog has been updated
- [ ] Documentation has been updated
- [ ] Unit tests pass locally (./scripts/test)
- [ ] Code is linted and styled (./scripts/format)

### CODE REVIEW

File: pctiler/pctiler/colormaps/__init__.py [hunk 1]
```diff
@@ -15,7 +16,7 @@
 # if we want documentation on par with the default RenderParams class
 ################################################################################
 registered_cmaps = cmap
-custom_colormaps: Dict[str, Dict[int, List[int]]] = {
+custom_colormaps: Dict[str, ColorMapType] = {
```
Comment (vincentsarago) [inline pctiler/pctiler/colormaps/__init__.py:19]: update to use new rio-tiler ColorMap Type

File: pctiler/pctiler/colormaps/jrc.py [hunk 2]
```diff
@@ -1,449 +1,451 @@
...
-        92: [39, 20, 236, 255],
-        93: [34, 17, 238, 255],
-        94: [29, 15, 240, 255],
-        95: [24, 12, 242, 255],
-        96: [19, 10, 245, 255],
-        97: [15, 7, 247, 255],
-        98: [10, 5, 250, 255],
-        99: [5, 2, 252, 255],
-        100: [0, 0, 255, 255],
+        0: (255, 255, 255, 255),
```
Comment (vincentsarago) [inline pctiler/pctiler/colormaps/jrc.py:7]: Type, is now a Tuple

File: pctiler/pctiler/endpoints/item.py [hunk 3]
```diff
@@ -40,151 +31,51 @@ class MapParams:
...
-            request: Request, map_params: MapParams = Depends(MapParams)
-        ) -> Response:
-
-            collection_id, item_id = (map_params.collection, map_params.item)
+pc_tile_factory = MultiBaseTilerFactory(
+    reader=ItemSTACReader,
+    path_dependency=ItemPathParams,
+    colormap_dependency=PCColorMapParams,
+    router_prefix=get_settings().item_endpoint_prefix,
+)
```
Comment (vincentsarago) [inline pctiler/pctiler/endpoints/item.py:39]: There is no more need to create a custom TilerFactory, we can use `MultiBaseTilerFactory` and set custom params

File: pctiler/pctiler/endpoints/item.py [hunk 4]
```diff
@@ -40,151 +31,51 @@ class MapParams:
...
-                    "itemUrl": item_url,
-                    "renderParams": get_param_str(render_config.render_params),
-                },
-            )
+    item_url = urljoin(
+        get_settings().stac_api_href,
+        f"collections/{collection}/items/{item}",
+    )
 
-    def register_poly_crop(self) -> None:
```
Comment (vincentsarago) [inline pctiler/pctiler/endpoints/item.py:117]: /crop is now in titiler tiler factory

File: pctiler/pctiler/endpoints/legend.py [hunk 5]
```diff
@@ -95,7 +96,18 @@ def make_colormap_image(
...
     """Use registered rio-tiler colormaps to create matplotlib colormap"""
+    cm = registered_cmaps.get(name)
+
+    # Make sure we can use `make_lut`
+    # see: https://github.com/cogeotiff/rio-tiler/blob/master/rio_tiler/colormap.py#L98-L108
+    if isinstance(cm, Sequence):
+        raise Exception("Cannot make a colormap from Intervals colormap")
+
+    if len(cm) > 256 or max(cm) >= 256:
+        raise Exception("Cannot make a colormap for discrete colormap")
```
Comment (vincentsarago) [inline pctiler/pctiler/endpoints/legend.py:107]: `make_lut` can only work with GDAL like colormap

File: pctiler/pctiler/endpoints/pg_mosaic.py [hunk 6]
```diff
@@ -1,84 +1,23 @@
...
 
 
-def AdditionalTileParam(
-    collection: str = Query(None, description="JSON encoded custom Colormap"),
-) -> Dict[str, Any]:
-    return {"collection": collection}
+@dataclass
+class AssetsBidxExprParams(dependencies.AssetsBidxExprParams):
 
+    collection: str = Query(None, description="STAC Collection ID")
```
Comment (vincentsarago) [inline pctiler/pctiler/endpoints/pg_mosaic.py:15]: we add `collection` to `AssetsBidxExprParams` params which will be used in all endpoints

File: pctiler/pctiler/endpoints/pg_mosaic.py [hunk 7]
```diff
@@ -1,84 +1,23 @@
...
-                    return dict(zip(fields, r))
-
-            return with_retry_connection(pool, get_search_info)
-
-
-pgstac_mosaic_factory = PCDynamicMosaicTilerFactory(
-    reader=CustomPGSTACBackend,  # type:ignore
-    router_prefix=get_settings().mosaic_endpoint_prefix,
+pgstac_mosaic_factory = MosaicTilerFactory(
+    reader=PGSTACBackend,
```
Comment (vincentsarago) [inline pctiler/pctiler/endpoints/pg_mosaic.py:19]: no more need for custom `MosaicTilerFactory` (no more retry connection)

File: pctiler/pctiler/reader.py [hunk 8]
```diff
@@ -43,8 +41,8 @@ def _get_asset_url(self, asset: str) -> str:
 
 
 @attr.s
-class MosaicSTACReader(MultiBaseReader):
-    """Simplified STAC Reader.
+class CustomSTACReader(pgstac_mosaic.CustomSTACReader):
```
Comment (vincentsarago) [inline pctiler/pctiler/reader.py:44]: we can extent `pgstac_mosaic.CustomSTACReader` instead of creating a new custom one

**Comment by vincentsarago** [comment]
Breaking changes, are described in https://github.com/developmentseed/titiler/discussions/396. TiTiler is trying to use better query parameters format, so moving away of `,` coma delimited list (`bidx=1,2,3` -> `bidx=1&bidx=2&bidx=3`)

For the Tiler API here are the breaking changes:
- `resampling_method` -> `resampling`
- `assets=b1,b2` -> `assets=b1&assets=b2` (for MultiBaseTilerFactory)
- **new** `asset_bidx=image|1,2,3` 
- **new** `asset_expression=image|b1+b2`

**Comment by mmcfarland** [comment]
@vincentsarago Sorry I didn't see that this PR was waiting for us to approve a CI build. I just kicked that off, and you can see the build failure in the action logs.

> pctiler/pctiler/reader.py:108: error: Name 'pgstac_mosaic' is not defined

**Comment by mmcfarland** [comment]
Getting failures on [preview endpoint](http://localhost:8080/data/item/preview.png?collection=naip&item=fl_m_2708064_sw_17_060_20191215_20200113&assets=image&bidx=1):

```
tiler_1     |   File "/usr/local/lib/python3.8/site-packages/titiler/core/factory.py", line 745, in preview
tiler_1     |     data = src_dst.preview(
tiler_1     |   File "/usr/local/lib/python3.8/site-packages/rio_tiler/io/base.py", line 651, in preview
tiler_1     |     output = multi_arrays(assets, _reader, **kwargs)
tiler_1     |   File "/usr/local/lib/python3.8/site-packages/rio_tiler/tasks.py", line 71, in multi_arrays
tiler_1     |     [data for data, _ in filter_tasks(tasks, allowed_exceptions=allowed_exceptions)]
tiler_1     |   File "/usr/local/lib/python3.8/site-packages/rio_tiler/tasks.py", line 71, in <listcomp>
tiler_1     |     [data for data, _ in filter_tasks(tasks, allowed_exceptions=allowed_exceptions)]
tiler_1     |   File "/usr/local/lib/python3.8/site-packages/rio_tiler/tasks.py", line 34, in filter_tasks
tiler_1     |     yield future.result(), asset
tiler_1     |   File "/usr/local/lib/python3.8/concurrent/futures/_base.py", line 432, in result
tiler_1     |     return self.__get_result()
tiler_1     |   File "/usr/local/lib/python3.8/concurrent/futures/_base.py", line 388, in __get_result
tiler_1     |     raise self._exception
tiler_1     |   File "/usr/local/lib/python3.8/concurrent/futures/thread.py", line 57, in run
tiler_1     |     result = self.fn(*self.args, **self.kwargs)
tiler_1     |   File "/usr/local/lib/python3.8/site-packages/rio_tiler/io/base.py", line 643, in _reader
tiler_1     |     data = cog.preview(
tiler_1     |   File "/opt/src/pctiler/pctiler/reader_cog.py", line 227, in preview
tiler_1     |     if "goeseuwest.blob.core.windows.net" in self.filepath:
tiler_1     | AttributeError: 'CustomCOGReader' object has no attribute 'filepath'
```

Failure for [mosaic/tile endpoint](http://localhost:8080/data/mosaic/tiles/a4f553c660d33c0ef46abefd279d5d07/WebMercatorQuad/11/360/769@2x?assets=elevation&colormap_name=terrain&collection=nasadem&format=png)

```
tiler_1     |   File "/usr/local/lib/python3.8/site-packages/titiler/pgstac/factory.py", line 148, in tile
tiler_1     |     data, _ = src_dst.tile(
tiler_1     |   File "/opt/src/pctiler/pctiler/reader.py", line 194, in tile
tiler_1     |     mosaic_assets = self.assets_for_tile(
tiler_1     |   File "/opt/src/pctiler/pctiler/reader.py", line 132, in assets_for_tile
tiler_1     |     return self.get_assets(Polygon.from_bounds(*bbox), **kwargs)
tiler_1     |   File "/usr/local/lib/python3.8/site-packages/cachetools/decorators.py", line 17, in wrapper
tiler_1     |     k = key(*args, **kwargs)
tiler_1     |   File "/opt/src/pctiler/pctiler/reader.py", line 136, in <lambda>
tiler_1     |     key=lambda self, geom, **kwargs: hashkey(self.path, str(geom), **kwargs),
tiler_1     | AttributeError: 'PGSTACBackend' object has no attribute 'path'
```

/register endpoint seems to be working ok. Let me know if I can help debug these.

**Comment by mmcfarland** [comment]
@vincentsarago things are generally working with the upgrade, though I've got a few test cases that are failing that I'm hoping are just related to a misunderstanding of the new TiTiler url params. Can you offer some suggestions here:

#### Previous working URL params
`item/preview.png?collection=naip&items=fl_m_2708064_sw_17_060_20191215_20200113&assets=image&bidx=1,2,3`

#### Not working
- `item/preview.png?collection=naip&item=fl_m_2708064_sw_17_060_20191215_20200113&assets=image&bidx=1&bidx=2&bidx=3`
- `{"detail":"PNG driver doesn't support 5 bands.  Must be 1 (grey), 2 (grey+alpha), 3 (rgb) or 4 (rgba) bands. "}`

Also tried `asset_bidx`
- `item/preview.png?collection=naip&item=fl_m_2708064_sw_17_060_20191215_20200113&asset_bidx=image|1,2,3`
- `{"detail":"assets must be defined either via expression or assets options."}`

The tile version also fails with a different error
- `mosaic/tiles/c1e2f5aeaa27eaa1f6009ea21b0d60cc/WebMercatorQuad/12/969/1699@2x?assets=image&bidx=1&bidx=2&bidx=3&collection=naip&format=png`
- `{"detail":"operands could not be broadcast together with shapes (4,512,512) (3,512,512) "}`

I'll note that _some_ NAIP / bidx request are succeeding, such as
- `item/preview.png?collection=naip&item=mi_m_4208544_sw_16_h_20160727&assets=image&bidx=1&bidx=2&bidx=3`

so I don't think it's URL params, necessarily. The requests that fail to render on this branch succeed for the same item/asset/bands in our staging deployment. Something to do with the COGs themselves?

Looks like other collections with requests involving bands fail similarly
- `item/preview.png?collection=aster-l1t&item=AST_L1T_00308152006170202_20150515181209&assets=SWIR&bidx=1,3,5&nodata=0`
- `{"detail":"PNG driver doesn't support 7 bands.  Must be 1 (grey), 2 (grey+alpha), 3 (rgb) or 4 (rgba) bands. "}`


All the other PC collections are rendering items correctly.

**Comment by vincentsarago** [comment]
thanks for the feedback @mmcfarland 
```
# before 
assets=image&bidx=1,2,3

# now
assets=image&asset_bidx=image|1,2,3
```

I realize now that rio-tiler/titiler could be smarter and maybe only accept `asset_bidx=image|1,2,3`

about the tile requests that succeed: 

when passing `bidx=1&bidx=2&bidx=3` this won't have any effects and thus fallback to the asset internal bands. It seems that you may have assets with 3 or 4 bands within the naip collection 🤷‍♂️

**Comment by mmcfarland** [comment]
Ah, makes sense. I should have taken the error message at its word 😬.

I fixed up the `asset_bidx` and `asset_expression` usage and it's all good now, thanks for help.

**Review by lossyrob — APPROVED** [review]

**Review by mmcfarland — APPROVED** [review]
Tested and working

**Pull request merged**